### ISSUE TITLE
Livechat - Office hours not being properly applied

### ISSUE THREAD

**Issue by dcbell** [RocketChat/Rocket.Chat#4898]
Your Rocket.Chat version: 0.45.0

We installed Rocket.Chat using snap. Things were looking good. But it looks like since DST went into effect, its not properly dealing with office hours for our Livechat. I have our office hours set from 9:00AM - 5:00PM. This morning at 8:30 AM I happened to notice the chat was live on the website. I went and looked and the hours had shifted to 8:00AM - 4:00PM. I assume because of the DST shift. I went ahead and bumped them all back to 9-5. I checked the system time and it was showing 8:48AM. But the website still shows chat available. Any suggestions?

![capture](https://cloud.githubusercontent.com/assets/930515/20103759/d538fcfe-a590-11e6-9612-1578b4a7cf62.PNG)

**Comment by marceloschmidt** [comment]
@dcbell the chat will always be visible on your website. What this setting does is that it will show all agents as offline until that time setting is reached. Is that not working for you?

**Comment by dcbell** [comment]
I totally get that part. But at 8:30 it wasn't greyed out with the away message. It was red and accepting chats.

**Comment by AlecTroemel** [comment]
When it is the closing time, it simply switches all livechat agents to `Not available`. Agents can still set themselves to available again after leaving office hours. It is more like a safeguard against people leaving the office with chat still up.. Perhaps a stronger setting of office hours could be created always show the offline chat form when out of office hours.

**Comment by designgurudotorg** [comment]
Where is the default time zone set?

**Comment by AlecTroemel** [comment]
@designgurudotorg All times are translated to Coordinated Universal Time (UTC) when passed to the server side

**Comment by designgurudotorg** [comment]
Thanks @AlecTroemel - hadn't seen that documented anywhere... so RC uses UTC by default

**Comment by AlecTroemel** [comment]
@designgurudotorg from the app (client side) you use and see your local timezone, which is determined by your browser. When you save, rocket chat converts those times to UTC under the hood for saving in the database. 

for example:
```
your timezone is EST
my timezone is CT
you set a closing time to 5pm EST
I will see it as 4pm CT
in the database that closing time will be 9pm UTC
```
Hope that clears some things up. I agree that its not documented well (not really documented at all really 😅 )

**Comment by designgurudotorg** [comment]
Oh, not sure if there's a bug in this then - I noticed that when I save the times I want (i.e. local to me - in EST) they seem to be off by a few hours (i.e. livechat automatically turns off earlier then it should)

Will check again this evening and report back.

q./

**Comment by AlecTroemel** [comment]
I'll try to look into it today as well!

**Comment by cdbessig** [comment]
I agree, we have this set to close at 4:30pm.  It is now 5:47pm and its still live on our site.

**Comment by designgurudotorg** [comment]
Apparently these are UTC by default

**Comment by msypes** [comment]
I'm not seeing this feature as having any effect at at all. For example, on a Monday, I unchecked Tuesday. The following morning, potential guests are presented with the usual prechat form or can start a new chat if they've previously filled that out. 
There is no indication that the service is outside of office hours. 
An agent who was already logged into the server still shows up on the backend as available, and can pick up the chat. An agent who logs on after the chat is initiated cannot even see the chat in the queue, although I think this bug has nothing to do with office hours.
The above conversation also doesn't make it clear where times are entered in local time or UTC.
Is there any accurate documentation on how office hours are supposed to work someone could point me to?

**Comment by uglimugli** [comment]
Not sure if this belongs to this topic but, office hours for us shows up twice

![office hours](https://user-images.githubusercontent.com/15074905/38766036-b5913ada-3fd4-11e8-8d52-289cec837604.png)
Snap install.

**Comment by dennisverspuij** [comment]
I agree with @msypes. When a user enables livechat during enabled office hours, and then the end of the hours are reached the user stays online for livechat, the hours setting do not seem to have any effect. Office days do however, when I disable the day the chat widget shows we're not online.

**Comment by cdbessig** [comment]
Hello,

Thank you for your email. Please note, I am out of the office from 5/25 - 6/2 and will return on Monday, June 4th, 2018. It should be assumed that I will have ZERO access to email and my responses will be delayed until the week of June 4th.  I assume it may take me a few days to catch up and respond to all emails, especially actionable ones.  

If you require an immediate response, please reach out to the general support email. Otherwise, I will answer your request as soon as possible when I return.  Kevin Blackwell and Vincent Oliver will be covering Project Management duties for ACTIVE in-progress projects.

General Support / Technical Issues
support@creatingdigital.com
Phone 201-475-1288

Vincent Oliver
voliver@creatingdigital.com
Phone 201-475-1288

Kevin Blackwell
kblackwell@creatingdigital.com
Phone: 201-475-1288

Thank you for your understanding.

**Comment by renatobecker** [comment]
Hello everyone!

Sorry for the late answer...
I just want to say that we found some issues related to the `Office Hours` feature.
We're going to check what's going on, I'll respond here when we have more information on that.

Thanks.

**Comment by EdwinFdoGarcia** [comment]
Livechat - Office hours duplication of days and impossible to change the hours of attention.



- Version of Rocket.Chat Server:  snap install rocketchat-server - Version | 0.74.2
Apps Engine Version 1.4.0
- Operating System: AWS EC2 t3.medium, Linux Ubuntu 16  x64
- Number of Running Instances:  1
- DB Replicaset Oplog: 
- NodeJS Version: v8.11.3

[https://drive.google.com/file/d/1FTvJDqlA2qwl82JjIG42qKPC7wLFViQQ/view?usp=sharing
<img width="1347" alt="captura de pantalla 2019-02-24 a la s 9 29 59 p m" src="https://user-images.githubusercontent.com/35706123/53311111-9138d880-387d-11e9-86c8-8ae968ffb7b0.png">
](url)

**Comment by renatobecker** [comment]
Hi @EdwinFdoGarcia!
Can you please check how many documents you have inside the `rocketchat_livechat_office_hour`?
Thanks.

**Comment by cdbessig** [comment]
Hello,

Thank you for your email. Please note, I have gone home sick for the day of 2/25.  As such, you should assume I will have ZERO access to email. I will begin to process my inbox and actionable items upon returning.  

If you have an URGENT issue and require an immediate response, please reach out to the general support email. Otherwise, I will answer your request as soon as possible when I return.  

General Support / Technical Issues
support@creatingdigital.com
Phone 201-475-1288



Thank you for your understanding.

**Comment by EdwinFdoGarcia** [comment]
Hi @renatobecker , thanks for your help. Sorry but I do not know where I can see rocketchat_livechat_office_hour. :(

**Comment by cdbessig** [comment]
Hello,

Thank you for your email. Please note, I am out of the country on business.  As such, you should assume I will have limited access to email.

For a faster response, please reach out to the general support email or Lori Talarico who is covering Account Management. Otherwise, I will answer your request as soon as possible which may be 1-2 business days.  

General Support / Technical Issues
support@creatingdigital.com
Phone 201-475-1288

Lori Talarico
ltalarico@creatingdigital.com
Phone: 201-475-1288

Thank you,

Brian

**Comment by renatobecker** [comment]
Hi @EdwinFdoGarcia, you just need to check the collection inside your database.

**Comment by r1der23** [comment]
if somebody needs, i solved like that

1) remove local tz to utc conversion in livechat admin panel
2) remove isClosingTimes check, because we don't need that. 
3) change isOpeningHours, replace isSame to isBetween and not every 10 sec script checks that current local time in range.
Now no problem with utc. i Have gmt+3 and if i need set 8am-11pm it was not possible, because it made 11am - 2am(next day)

![image](https://user-images.githubusercontent.com/11757905/59983859-c509ec00-962c-11e9-8609-060bd219f535.png)
![image](https://user-images.githubusercontent.com/11757905/59983864-cd622700-962c-11e9-8220-e47d1cd2c432.png)
![image](https://user-images.githubusercontent.com/11757905/59983869-d94de900-962c-11e9-8a04-c235e0a0b6ff.png)

download files
https://www.sendspace.com/filegroup/QNgSDVE55ID9fPwQwrz8DABzrETrwHhB

if it helps you can contribute that, i don't know how

**Comment by KevLehman** [comment]
This feature was re-created in recent releases, feel free to open a new issue if the issue persist.
